modules/core/src/main/java/org/apache/ignite/internal/binary/streams/BinaryMemoryAllocator.java

```diff
@@ -16,41 +16,258 @@
  */
-public final class BinaryMemoryAllocator {
+public abstract class BinaryMemoryAllocator {
+    /** Buffer size re-check frequency. */
+    private static final Long CHECK_FREQ = Long.getLong(IGNITE_MARSHAL_BUFFERS_RECHECK, 10000);
+
+    /** */
+    private static final int POOL_SIZE = Integer.getInteger(IGNITE_MARSHAL_BUFFERS_PER_THREAD_POOL_SIZE, 32);
+
     /** Memory allocator instance. */
-    public static final BinaryMemoryAllocator INSTANCE = new BinaryMemoryAllocator();
+    public static final BinaryMemoryAllocator INSTANCE = new ThreadLocalAllocator();
 
-    /** Holders. */
-    private static final ThreadLocal<BinaryMemoryAllocatorChunk> holders = new ThreadLocal<>();
+    /** Memory allocator instance. */
+    public static final BinaryMemoryAllocator POOLED = new PooledAllocator();
 
-    /**
-     * Ensures singleton.
-     */
-    private BinaryMemoryAllocator() {
-        // No-op.
-    }
+    /** */
+    public abstract BinaryMemoryAllocatorChunk chunk();
+
+    /** */
+    public abstract boolean isAcquired();
+
+    /** */
+    private static class ThreadLocalAllocator extends BinaryMemoryAllocator {
+        /** Holders. */
+        private final ThreadLocal<BinaryMemoryAllocatorChunk> holders = new ThreadLocal<>();
+
+        /** {@inheritDoc} */
+        @Override public BinaryMemoryAllocatorChunk chunk() {
+            BinaryMemoryAllocatorChunk holder = holders.get();
+
+            if (holder == null)
+                holders.set(holder = new Chunk());
+
+            return holder;
+        }
+
+        /**
+         * Checks whether a thread-local array is acquired or not.
+         * The function is used by Unit tests.
+         *
+         * @return {@code true} if acquired {@code false} otherwise.
+         */
+        @Override public boolean isAcquired() {
+            BinaryMemoryAllocatorChunk holder = holders.get();
+
+            return holder != null && holder.isAcquired();
+        }
+
+        /**
+         * Memory allocator chunk.
+         */
+        private static class Chunk implements BinaryMemoryAllocatorChunk {
+            /** Data array */
+            private byte[] data;
+
+            /** Max message size detected between checks. */
+            private int maxMsgSize;
+
+            /** Last time array size is checked. */
+            private long lastCheckNanos = System.nanoTime();
+
+            /** Whether the holder is acquired or not. */
+            private boolean acquired;
+
+            /** {@inheritDoc} */
+            @Override public byte[] allocate(int size) {
+                if (acquired)
+                    return new byte[size];
+
+                acquired = true;
+
+                if (data == null || size > data.length)
+                    data = new byte[size];
+
+                return data;
+            }
+
+            /** {@inheritDoc} */
+            @Override public byte[] reallocate(byte[] data, int size) {
+                byte[] newData = new byte[size];
```

Comment:
Should not we check if the `size` fits `data` before allocating new memory?

Reply:
All allocations are checked by corresponding binary writer, the method won't be called if allocation isn't needed